### ISSUE TITLE
Cookbook compile stage error: uninitialized constant Docker::Network

### ISSUE THREAD

**Issue by trinitronx** [chef-cookbooks/docker#667]
Using updated docker cookbook, I get:

```
Starting Chef Client, version 12.5.1
resolving cookbooks for run list: ["chef-client::delete_validation", "parted", "apt", "build-essential", "ntp",  "docker"]
Synchronizing Cookbooks:
  - cron (1.4.0)
  - logrotate (1.6.0)
  - parted (2.1.2)
  - chef-client (3.6.0)
  - apt (2.9.2)
  - build-essential (2.0.4)
  - ntp (1.8.2)
  - docker (2.5.8)
Compiling Cookbooks...
WARN: Unresolved specs during Gem::Specification.reset:
      httpclient (>= 2.2.0.2, ~> 2.2)
WARN: Clearing out unresolved specs.
Please report a bug if this causes problems.

================================================================================
Recipe Compile Error in /var/chef/cache/cookbooks/docker/libraries/docker_network.rb
================================================================================

NameError
---------
uninitialized constant Docker::Network

Cookbook Trace:
---------------
  /var/chef/cache/cookbooks/docker/libraries/docker_network.rb:18:in `<class:DockerNetwork>'
  /var/chef/cache/cookbooks/docker/libraries/docker_network.rb:2:in `<module:DockerCookbook>'
  /var/chef/cache/cookbooks/docker/libraries/docker_network.rb:1:in `<top (required)>'

Relevant File Content:
----------------------
/var/chef/cache/cookbooks/docker/libraries/docker_network.rb:

 11:      property :driver, String
 12:      property :driver_opts, PartialHashType
 13:      property :gateway, [String, Array, nil], coerce: proc { |v| coerce_gateway(v) }
 14:      property :host, [String, nil], default: lazy { default_host }, desired_state: false
 15:      property :id, String
 16:      property :ip_range, [String, Array, nil], coerce: proc { |v| coerce_ip_range(v) }
 17:      property :ipam_driver, String
 18>>     property :network, Docker::Network, desired_state: false
 19:      property :network_name, String, name_property: true
 20:      property :subnet, [String, Array, nil], coerce: proc { |v| coerce_subnet(v) }
 21:
 22:      alias aux_address auxiliary_addresses
 23:
 24:      load_current_value do
 25:        begin
 26:          with_retries { network Docker::Network.get(network_name, {}, connection) }
 27:        rescue Docker::Error::NotFoundError


Running handlers:
[2016-02-28T02:16:21-07:00] ERROR: Running exception handlers
Running handlers complete
[2016-02-28T02:16:21-07:00] ERROR: Exception handlers complete
Chef Client failed. 0 resources would have been updated
[2016-02-28T02:16:21-07:00] FATAL: Stacktrace dumped to /var/chef/cache/chef-stacktrace.out
[2016-02-28T02:16:21-07:00] ERROR: uninitialized constant Docker::Network
[2016-02-28T02:16:21-07:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
```

Seems that this cookbook may depend on [classes defined in `docker-api`][1]?  Looks like it tries to use `_autoload.rb` to load it, but for some reason does not work?  I haven't seen `_autoload` used before, is this a special reserved name in Chef now?

Tested with:
```
Chef Development Kit Version: 0.10.0
chef-client version: 12.5.1
berks version: 4.0.1
kitchen version: 1.4.2
```

[1]: http://www.rubydoc.info/gems/docker-api/1.26.0/Docker/Network

**Comment by someara** [comment]
`_autoload` is not reserved.. it's named that so comes first alphabetically (ruby load order).

I noticed you're missing compat_resource in the 'Synchronizing Cookbooks' section, which strikes me as odd... you should have '~> 12.7.1' in there, as it's a dependency in the metadata.rb

Do you have a chef-server in there? 'chef-client::delete_validation' implies that you do. What version is it?

-s

**Comment by trinitronx** [comment]
yes, there is a Chef server involved here.  I'm noticing that I did have `compat_resource (12.7.3)` and **many** other cookbooks in the 'Synchronizing Cookbooks' section due to the run list in that test, but I remember trimming some of the output to try and make it more readable.  I noticed that I had extra cookbooks in there which I didn't think were related to the problem, so I tried to avoid spamming that giant block of text.

Anyway,  I tried re-running today and ended up with a different error, I'm not sure I'm using this cookbook correctly

```
================================================================================
  Recipe Compile Error
  ================================================================================

  Chef::Exceptions::RecipeNotFound
  --------------------------------
  could not find recipe default for cookbook docker
```

I'm not sure what I did differently, and haven't had the time to debug further due to this being a home project & having a full-time job focused on other things at the moment.  I'm not sure I'm using this cookbook correctly anymore (I used to use `v0.37` and decided it's about time to take the upgrade plunge).  However, this cookbook has indeed followed SemVer and does have a lot of breaking & behavioral changes since then.  I just double checked and realized there are no files under `recipes/` anymore.  Usually with LWRP cookbooks, I'm used to seeing at least a blank `default.rb` in there.

Am I missing something new with Chef 12 or how to use this cookbook?

**Comment by chasebolt** [comment]
This cookbook is purely library driven and you do not need to include any recipe to start using it. Looks like you are trying to `include_recipe 'docker'`, which isn't necessary. All you need to use this cookbook is `depends 'docker'` in your metadata and start using the examples from the README.

**Comment by trinitronx** [comment]
@chasebolt :  Thanks for the explanation.  I was able to find some more time to look more closely at all the stuff in the CHANGELOG and git diffs from way back.  I realized that I was simply too tired to remember the wrapper cookbook pattern when I first found personal time at night looking at this.

I was able to get this working again using a recipe like:

```ruby
include_recipe 'apt'
include_recipe 'apt-docker'
include_recipe 'docker'

docker_installation_package 'default' do
  action :create
end
```

And adding to `metadata.rb`:

```ruby
depends 'apt'
depends 'apt-docker', '~> 0.3.0'
depends 'docker', '~> 2.5.8'
```
